File: PULSE.Services/Data/Abstract/Product.cs
namespace PULSE.Services.Data
{
    public class Product : AuditableEntity
    {
        public Product()
        {
            OrderDetails = new HashSet<OrderDetail>();
            Images = new HashSet<ProductImage>();
            ServicingParts = new HashSet<ServicingPart>();
        }

        public int? BrandId { get; set; }
        public int? ProductCategoryId { get; set; }
        public string? ProductNumber { get; set; }
        public string? Model { get; set; }
        public string? Description { get; set; }
        public decimal? Price { get; set; }
        public string? Discriminator { get; set; }
        public byte[]? ThumbImage { get; set; }

        public virtual Brand? Brand { get; set; }
        public virtual ICollection<OrderDetail> OrderDetails { get; set; }
        public virtual ICollection<ProductImage> Images { get; set; }
        public virtual ICollection<ServicingPart> ServicingParts { get; set; }
    }
}

File: pulse_mobile/lib/model/abstract/product.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pulse_mobile/model/product_category/product_category.dart';
import 'package:pulse_mobile/model/product_image/product_image.dart';

import '../brand/brand.dart';
import 'auditable_model.dart';

part 'product.g.dart';

@JsonSerializable()
class Product extends AuditableModel {
  int? brandId;
  int? productCategoryId;
  String? productNumber;
  String? model;
  String? description;
  String? discriminator;
  ProductCategory? productCategory;
  Brand? brand;
  double? price;
  List<ProductImage>? images;

  Product();

  factory Product.fromJson(Map<String, dynamic> json) =>
      _$ProductFromJson(json);

  Map<String, dynamic> toJson() => _$ProductToJson(this);

  @override
  String toString() {
    return "$brand $model";
  }
}

enum ProductType { Bicycle, Gear, Part }
